components/script/dom/node.rs

```diff
@@ -471,19 +472,27 @@ impl Node {
     }
 
     /// Iterates over this node and all its descendants, in preorder.
-    pub fn traverse_preorder(&self) -> TreeIterator {
-        TreeIterator::new(self)
+    pub fn traverse_preorder<T>(&self) -> TreeIterator<T>
+        where T: DerivedFrom<Node>
+    {
+        TreeIterator::<T>::new(self)
     }
 
-    pub fn inclusively_following_siblings(&self) -> NodeSiblingIterator {
-        NodeSiblingIterator {
-            current: Some(Root::from_ref(self)),
+    pub fn inclusively_following_siblings<T>(&self) -> SiblingIterator<T>
+        where T: DerivedFrom<Node>
+    {
+        SiblingIterator::<T> {
+            current: Root::downcast::<T>(Root::from_ref(self)),
```

Comment:
This is incorrect, the node following `self` may be a `T`, but if `self` is not a `T`, `current` will be set to `None`.

You probably want:

```rust
SiblingIterator {
    current: Root::downcast(Root::from_ref(self)).or_else(|| get_next_sibling(self)),
    phantom: PhantomData,
}
```